doc/java.md

```diff
@@ -0,0 +1,298 @@
+    } else {
+      if (lt# [Int] anf$5 (7 : Int)) {
+        keep_fact_in_history;
+      }
+
+    }
+
+  }
+  for_facts (gen$fact : Int) in new {
+    let anf$16 = snd# [Int] [DateTime] gen$fact;
+    let anf$17 = DateTime_daysDifference# anf$16 gen$date;
+    let anf$18 = le# [Int] anf$17 (14 : Int);
+    let anf$20 = ge# [Int] anf$17 (7 : Int);
+    if (and# anf$18 anf$20) {
+      read acc$conv$4 = acc$conv$4;
+      write acc$conv$4 = add# acc$conv$4 (1 : Int);
+    } else {
+      if (lt# [Int] anf$17 (7 : Int)) {
+        keep_fact_in_history;
+      }
+
+    }
+
+  }
+  read conv$4 = acc$conv$4;
+  return conv$4;
+}
+```
+
+Java:
+```
+void Compute(IcicleMagic icicle)
+{
+    // gen$date = DATE
+    DateTime gen_date = icicle.SnapshotDate();
+
+    // init acc$conv$4 = 0 : Int (Windowed);
+    int acc_conv_4 = 0;
+
+    // for_facts (gen$fact : Int) in history
+    icicle.StartHistory();
+    while (icicle.HasField()) {
+        // let anf$4 = snd# [Int] [DateTime] gen$fact;
+        DateTime anf_4 = icicle.CurrentTime();
+        // let anf$5 = DateTime_daysDifference# anf$4 gen$date;
+        int anf_5 = DateTime.daysDifference(anf_4, gen_date);
+
+        // let anf$6 = le# [Int] anf$5 (14 : Int);
+        bool anf_6 = anf_5 <= 14;
+        // let anf$8 = ge# [Int] anf$5 (7 : Int);
+        bool anf_8 = anf_8 >= 7;
+
+        // if (and# anf$6 anf$8)
+        if (anf_6 && anf_8) {
+            // read acc$conv$4 = acc$conv$4;
+            int acc_conv_4_read = acc_conv_4;
+            // write acc$conv$4 = add# acc$conv$4 (1 : Int);
+            acc_conv_4 = acc_conv_4_read + 1;
+        } else {
+            // if (lt# [Int] anf$5 (7 : Int))
+            if (anf_5 < 7) {
+                // keep_fact_in_history;
+                icicle.KeepFactInHistory();
+            }
+        }
+    }
+
+    // for_facts (gen$fact : Int) in new
+    icicle.StartNew();
+    while (icicle.HasField()) {
+        // let anf$16 = snd# [Int] [DateTime] gen$fact;
+        DateTime anf_16 = icicle.CurrentTime();
+        // let anf$17 = DateTime_daysDifference# anf$16 gen$date;
+        int anf_17 = DateTime.daysDifference(anf_16, gen_date);
+
+        // let anf$18 = le# [Int] anf$17 (14 : Int);
+        bool anf_18 = anf_17 <= 14;
+        // let anf$20 = ge# [Int] anf$17 (7 : Int);
+        bool anf_20 = anf_17 >= 7;
+
+        // if (and# anf$18 anf$20)
+        if (anf_18 && anf_20) {
+            // read acc$conv$4 = acc$conv$4;
+            int acc_conv_4_read = acc_conv_4;
+            // write acc$conv$4 = add# acc$conv$4 (1 : Int);
+            acc_conv_4 = acc_conv_4_read + 1;
+        } else {
+            // if (lt# [Int] anf$17 (7 : Int))
+            if (anf_17 < 7) {
+                icicle.KeepFactInHistory();
+            }
+        }
+    }
+
+    // read conv$4 = acc$conv$4;
+    int conv_4 = acc_conv_4
+    // return conv$4;
+    icicle.Output<Integer>("CountLastWeek", new Integer(conv_4));
```

Comment:
> I'm guessing accessing an int field from an ivory struct will allocate

The values from the structs are allocated when we deserialise from thrift, so by the time it reaches icicle there's no allocations just to read the value, only to write back again. I think what you have is fine - and matches what we have at the moment.